compose.yml

```diff
@@ -0,0 +1,109 @@
+services:
+  www:
+    # Builds the Docker image from scratch
+    build:
+      dockerfile: docker/caddy/Dockerfile
+    container_name: mbin-www
+    restart: unless-stopped
+    networks:
+      - mbin_external_network
+    command: caddy run --config /etc/caddy/Caddyfile --adapter caddyfile
+    healthcheck:
+      test: ["CMD-SHELL", "wget -O /dev/null http://localhost:2019/metrics || exit 1"]
+    env_file:
+      - .env
+    depends_on:
+      - php
+
+  php:
+    # Builds the Docker image from scratch
+    build:
+      dockerfile: docker/php/Dockerfile
+    container_name: mbin-php
+    command: php-fpm
+    working_dir: /var/www/mbin
+    restart: unless-stopped
+    networks:
+      - mbin_external_network
+    healthcheck:
+      test:
+        [
+          "CMD-SHELL",
+          "REQUEST_METHOD=GET SCRIPT_NAME=/ping SCRIPT_FILENAME=/ping cgi-fcgi -bind -connect localhost:9000 || exit 1",
+        ]
+    env_file:
+      - .env
+    depends_on:
+      redis:
+        condition: service_healthy
+      db:
+        condition: service_healthy
+      rabbitmq:
+        condition: service_healthy
+      messenger:
+        # The main application depends on being able to send messages
+        condition: service_healthy
+
+  messenger:
+    # Builds the Docker image from scratch
+    build:
+      dockerfile: docker/php/Dockerfile
+    restart: unless-stopped
+    networks:
+      - mbin_external_network
+    command: bin/console messenger:consume --all --time-limit=3600
+    working_dir: /var/www/mbin
+    healthcheck:
+      test: ["CMD-SHELL", "ps aux | grep 'messenger[:]consume' || exit 1"]
+    env_file:
+      - .env
+    depends_on:
+      - redis
+      - db
+      - rabbitmq
+
+  redis:
+    image: redis:alpine
+    container_name: mbin-redis
+    restart: unless-stopped
+    networks:
+      - mbin_external_network
+    command: /bin/sh -c "redis-server --requirepass $${REDIS_PASSWORD}"
+    env_file:
+      - .env
+    healthcheck:
+      test: ["CMD", "redis-cli", "ping"]
+
+  db:
+    image: postgres:${POSTGRES_VERSION:-16}-alpine
+    container_name: mbin-db
+    restart: unless-stopped
+    networks:
+      - mbin_external_network
```

Comment:
DB & Redis can definitely be an internal network  (`internal: true`). Yes, you can have multiple networks in Docker. And in Compose you can also add multiple networks to a single service (eg. messenger can both have internal and external docker networks).

Reply:
:thinking: I'm not sure what advantage that would have for the `php` and `messenger` services. If I'm not mistaken those require outgoing access to external networks in order to download stuff and communicate with other instances. The `php` service should probably be able to access direct incoming traffic. Dunno about the messenger.

The DB and redis could probably be purely internal as I don't believe they'll ever need to access external resources.

Should this be done in another PR? I feel like it could expand the scope of this PR and it might be good to have a separate PR with one concern (docker compose networking) for posterity.

Reply:
As fair as I know only the www service (and maybe also the rabbitmq if you enabled the management interface) needs access to the external network, the rest of all the services can do internal communication between the services.

Sure, this can also be done in a follow-up PR.

Reply:
Created [a ticket](https://github.com/MbinOrg/mbin/issues/1073) so it won't be forgotten.

Reply:
I now get:

```
Error response from daemon: driver failed programming external connectivity on endpoint mbin-db (c1c3e1b40be94b691d7a32acbdcde8723b9a7402b68bbb64598c5533e354b1e5): failed to bind port 127.0.0.1:5432/tcp: Error starting userland proxy: listen tcp4 127.0.0.1:5432: bind: address already in use
```

So this external network will interfere with any host ports that run for example MySQL/MariaDB. Therefore, I reopened this conversation.